### ISSUE TITLE
Blocking query on health/service sometimes not returning on service change

### ISSUE THREAD

**Issue by plorenz** [hashicorp/consul#2814]
Running test locally with Docker. Using docker image derived from Hashicorp official image for 0.7.3, with just a custom start script added.

### `consul version` for both Client and Server
==> WARNING: BootstrapExpect Mode is specified as 1; this is the same as Bootstrap mode.
==> WARNING: Bootstrap mode enabled! Do not enable unless necessary
==> Starting Consul agent...
==> Starting Consul agent RPC...
==> Consul agent running!
           Version: 'v0.7.3'
           Node ID: 'a710c6c1-82ca-06b3-b206-1ca7ac605cec'
         Node name: 'plorenz-lt'
        Datacenter: 'dc1'
            Server: true (bootstrap: true)
       Client Addr: 0.0.0.0 (HTTP: 8500, HTTPS: -1, DNS: 8600, RPC: 8400)
      Cluster Addr: 192.168.3.60 (LAN: 8301, WAN: 8302)
    Gossip encrypt: false, RPC-TLS: false, TLS-Incoming: false
             Atlas: <disabled>

### `consul info` for both Client and Server
Server:
```
/ # consul info
agent:
        check_monitors = 1
        check_ttls = 0
        checks = 1
        services = 3
build:
        prerelease = 
        revision = 'a90bb8f
        version = 0.7.3
consul:
        bootstrap = true
        known_datacenters = 1
        leader = true
        leader_addr = 192.168.3.60:8300
        server = true
raft:
        applied_index = 590
        commit_index = 590
        fsm_pending = 0
        last_contact = never
        last_log_index = 590
        last_log_term = 2
        last_snapshot_index = 0
        last_snapshot_term = 0
        latest_configuration = [{Suffrage:Voter ID:192.168.3.60:8300 Address:192.168.3.60:8300}]
        latest_configuration_index = 1
        num_peers = 0
        protocol_version = 1
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Leader
        term = 2
runtime:
        arch = amd64
        cpu_count = 8
        goroutines = 67
        max_procs = 2
        os = linux
        version = go1.7.4
serf_lan:
        encrypted = false
        event_queue = 1
        event_time = 2
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 1
        members = 1
        query_queue = 0
        query_time = 1
serf_wan:
        encrypted = false
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 1
        members = 1
        query_queue = 0
        query_time = 1

```
### Description of the Issue (and unexpected/desired result)
Testing watching a service come and go. Using docker, I start the service, stop, start, etc. Using curl to wait for service changes. At somepoint when the service goes away, the blocking query doesn't return until it times out. When it does return it, shows stale data and the same index that it was queried with.

Example:
This query was started. I then killed the service. The query didn't return for ~10 minutes. When it did return, it showed the service still active and an unchanged index. I ran a second query immediately after killing the service. It return immediately and properly showed the service as gone and an updated index.

First query
```
$ curl -v localhost:8500/v1/health/service/myservice\?index=316\&wait=3600s
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8500 (#0)
> GET /v1/health/service/myservice?index=316&wait=3600s HTTP/1.1
> Host: localhost:8500
> User-Agent: curl/7.50.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/json
< X-Consul-Index: 316
< X-Consul-Knownleader: true
< X-Consul-Lastcontact: 0
< Date: Wed, 22 Mar 2017 15:48:38 GMT
< Content-Length: 923
< 
* Connection #0 to host localhost left intact
[{"Node":{"ID":"a710c6c1-82ca-06b3-b206-1ca7ac605cec","Node":"plorenz-lt","Address":"192.168.3.60","TaggedAddresses":{"lan":"192.168.3.60","wan":"192.168.3.60"},"Meta":{},"CreateIndex":4,"ModifyIndex":8},"Service":{"ID":"myservice-9092","Service":"myservice","Tags":["containerid:cff775d9118415453371e267a3d1a639e0859e4c88777a70107ea871786a3663"],"Address":"","Port":9092,"EnableTagOverride":false,"CreateIndex":309,"ModifyIndex":309},"Checks":[{"Node":"plorenz-lt","CheckID":"serfHealth","Name":"Serf Health Status","Status":"passing","Notes":"","Output":"Agent alive and reachable","ServiceID":"","ServiceName":"","CreateIndex":4,"ModifyIndex":4},{"Node":"plorenz-lt","CheckID":"31b619aa-fa84-4597-bdce-540b7887bb39","Name":"31b619aa-fa84-4597-bdce-540b7887bb39","Status":"passing","Notes":"","Output":"TCP connect 192.168.3.60:9092: Success","ServiceID":"myservice-9092","ServiceName":"myservice","CreateIndex":310,"ModifyIndex":315}]}]
```

Second query:
```
$ curl -v localhost:8500/v1/health/service/myservice\?index=316\&wait=3600s
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8500 (#0)
> GET /v1/health/service/myservice?index=316&wait=3600s HTTP/1.1
> Host: localhost:8500
> User-Agent: curl/7.50.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/json
< X-Consul-Index: 324
< X-Consul-Knownleader: true
< X-Consul-Lastcontact: 0
< Date: Wed, 22 Mar 2017 15:38:58 GMT
< Content-Length: 2
< 
* Connection #0 to host localhost left intact
[]
```
I didn't see any fixes in 0.7.4 or 0.7.5 that looked related, but if you want, I can try to reproduce with a 0.7.5.

**Comment by slackpad** [comment]
Hi @plorenz thanks for opening an issue. I tried to repro this locally and can't seem to get it to fail in this way. If you can try 0.7.5 I'd appreciate it as the fix there was related to the blocking mechanism, though it doesn't _seem_ related to what you are seeing. Also, does it get stuck every time, or sporadically?

**Comment by plorenz** [comment]
@slackpad I'll try it with 0.7.5. It doesn't happen every time. It seems to take several cycles of register/unregister before it fails on an unregister. It also seems to mostly happen on a specific service which takes a few minutes to come up and also registers two services (one for SSL, one for non-SSL). I tried it on a different service which comes up quickly, hoping for a faster repro cycle, but no luck. 

I think once it happens it seems to happen more consistently. I'll try to verify that and see if it happens on each unregister after the first failure.

**Comment by plorenz** [comment]
@slackpad I was unable to reproduce the issue with 0.7.5, which is great. Because it was so hard to reproduce, we'll keep an eye for similar behavior in future. If does recur, we'll try to come up with more easily reproducible test case :)

**Comment by slackpad** [comment]
Hey @plorenz thank you for the follow up! That bug fix in 0.7.5 was pretty deep down, and could result in the wakeup accounting getting off, so it's definitely possible that was it. Thanks again!
